hcl2template/plugin.go

```diff
@@ -128,6 +128,14 @@ func (cfg *PackerConfig) initializeBlocks() hcl.Diagnostics {
 	var diags hcl.Diagnostics
 
 	for _, build := range cfg.Builds {
+		// Since the build's contents may not have been dynamically
+		// expanded when we first loaded the config from the template
+		// file, we decode it now.
+		diags = append(diags, build.finalizeDecode(cfg)...)
```

Comment:
I feel like there is some important context missing here. In the decodeFile you have a comment for the BuildBlock and SourceBlock cases that reads "The final decoding step (which requires an up-to-date context) will be done when we need it." but it is not clear form this line that the time is now.

Also correct me if I'm wrong but a call to finalilzeDecode requires that the buildBlock has been decoded at least once. So calling `build.finalizeDecode` on an zero value BuildBlock would do what? 

It might be my understanding but there feels like there is a dependency chain that is not captured in the comments or in the type itself.

Reply:
So the build block is not completely decoded at this point, we only partially decode its top-level contents, and we certainly didn't try to expand the potential `dynamic` blocks inside it.
Same for the `source` blocks.

Regarding this comment, it's something I did not change, and that was how things worked before too. This `initializeBlocks` function basically expands the dynamic blocks, and filling-in the blanks in the structures that Packer uses.
At that point we do the decoding for all those, like provisioners, post-processors, and the sources themselves.

That being said, the `may` indicates that we could try before, but in the current implementation, this is very much not the case, we only get basic information from the blocks, and we need to call the `finalizeDecode` function on all the build and source blocks.

I'll change the phrasing here to remove this confusion, but does my explanation clarifies things for you?

Reply:
Come to think of it, I feel that the confusion stems from the vocabulary not being clear, so I'll leave this:

* decode (parser): lift the configuration from the HCL file and start building the components for the Packer config. At this point, we have not evaluated anything, nor should we. This step's purpose is only to parse the HCL, check if it is valid, and that there are no unexpected blocks/attributes in the config (for those we do know in advance what the format is).
* evaluateVariables/executeDatasources: there, we decode (as gohcl.Decode) the configuration for the entity, and send it to the component responsible, which effectively executes it. For variables there's no execution here, only expressions to evaluate, hence the difference in terminology.
* finalizeDecode (later): there, we expand the dynamic blocks where applicable, and we do as for datasources, we decode (as in gohcl.Decode), and send those configs to the plugins for validation.
* build: finally, once everything passed the validation and we have `*packer.CoreBuild` objects, we can then execute the builds.

With this in mind, I think the `decode` name is not representative of what is done during parsing. What would you think of some alternatives like `lift` or `load`? This would make it less likely to have a conflict in naming, and the `finalizeDecode` functions can then becode `decode` only, as it's what they essentially do.

I struggle in finding an appropriate name for this step of preparation of the structures, so feel free to suggest anything else, I'm eager to see what you think may fit the purpose better.